Prefer the Redis unix socket if available

ekohl opened the pull request:
Unix sockets typically have lower overhead and also allows setting stricter permissions. While iptables can be used to limit access using users, file permissions are much easier to manage.

Includes https://github.com/theforeman/puppet-pulpcore/pull/218.

Currently a draft, but I'm exploring the option of deploying a dedicated Redis instance just for Pulp (if needed). This means Redis can be tuned different. Foreman needs Redis to be in persistent mode but Pulp doesn't always:

| Tasking | Caching | Redis mode |
|---------+---------+------------+
| RQ      | Yes     | Persistent |
| RQ      | No      | Persistent |
| PG      | Yes     | Empheral   |
| PG      | No      | Off        |

ehelms commented:
> Currently a draft, but I'm exploring the option of deploying a dedicated Redis instance just for Pulp (if needed). This means Redis can be tuned different. Foreman needs Redis to be in persistent mode but Pulp doesn't always:

Since Foreman may also want to do caching potentially with a Redis, would we be looking at:

 * 3 instances, one for Dynflow, one for Pulp cache, one for Foreman cache
 * 2 instances, one for persistent workloads (e.g. Dynflow) and one for caching (Foreman or Pulp)

And does that change any of the design?

ehelms commented on manifests/config.pp:
```diff
@@ -17,6 +17,15 @@
     ensure_newline => true,
   }
 
+  if $redis::unixsocket != '' {
```
What would be configuring this given we only `include redis` ?

ekohl replied:
Technically it could have been changed via hiera. Outside of the installer this can also happen directly via manifests.

ekohl commented:
> Since Foreman may also want to do caching potentially with a Redis, would we be looking at:
> 
>     * 3 instances, one for Dynflow, one for Pulp cache, one for Foreman cache
> 
>     * 2 instances, one for persistent workloads (e.g. Dynflow) and one for caching (Foreman or Pulp)
> 
> 
> And does that change any of the design?

Possibly. At this point I want to find out if we can even use a unix socket in the first place. We may run into problems like permissions, maybe something inside Pulp itself and probably SELinux policies.

At the same time, should start to investigate if Foreman and Dynflow can.

ekohl commented:
Rebased, but it's now split into two commits. First there's https://github.com/theforeman/puppet-pulpcore/pull/331 which makes the Redis URL configurable (which is useful for remote Redis servers). Then this expands it to use the unix socket if available.
